geoh5py/objects/surveys/electromagnetics/base.py

```diff
@@ -0,0 +1,415 @@
+
+            if len(value.properties) != len(self.channels):
+                raise ValueError(
+                    f"Number of properties in group '{value.name}' "
+                    + "differ from the number of 'channels'."
+                )
+
+            if value.name not in self.metadata["EM Dataset"]["Property groups"]:
+                self.metadata["EM Dataset"]["Property groups"].append(value.name)
+
+    @property
+    def input_type(self):
+        """Data input type. Must be one of 'Rx', 'Tx' or 'Tx and Rx'"""
+        if "Input type" in self.metadata["EM Dataset"]:
+            return self.metadata["EM Dataset"]["Input type"]
+
+        return None
+
+    @input_type.setter
+    def input_type(self, value: str):
+        if self.default_input_types is not None:
+
+            if value not in self.default_input_types:
+                raise ValueError(
+                    "Input 'input_type' must be one of "
+                    f"{self.default_input_types}. {value} provided."
+                )
+            self.edit_metadata({"Input type": value})
+
+    @property
+    def metadata(self) -> dict:
+        """Metadata attached to the entity."""
+        if getattr(self, "_metadata", None) is None:
+            metadata = self.workspace.fetch_metadata(self.uid)
+
+            if metadata is None:
+                metadata = self.default_metadata
+                for name in ["Receivers", "Transmitters"]:
+                    if name in type(self).__name__:
+                        metadata["EM Dataset"][name] = self.uid
+                self.metadata = metadata
+            else:
+                self._metadata = metadata
+
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, values: dict):
+        if not isinstance(values, dict):
+            raise TypeError("'metadata' must be of type 'dict'")
+
+        if "EM Dataset" not in values:
+            values = {"EM Dataset": values}
+
+        for key in self.default_metadata["EM Dataset"]:
+            if key not in values["EM Dataset"]:
+                raise KeyError(f"'{key}' argument missing from the input metadata.")
+
+        for key, value in values["EM Dataset"].items():
+            if key in ["Receivers", "Transmitters"] and isinstance(value, str):
+                values["EM Dataset"][key] = uuid.UUID(value)
+
+        self._metadata = values
+        self.modified_attributes = "metadata"
+
+    @property
+    def receivers(self) -> BaseEMSurvey | None:
+        """
+        The associated TEM receivers.
+        """
+        if getattr(self, "_receivers", None) is None:
+            if self.metadata is not None and "Receivers" in self.metadata["EM Dataset"]:
+                rx_uid = self.metadata["EM Dataset"]["Receivers"]
+
+                try:
+                    self._receivers = self.workspace.get_entity(rx_uid)[0]
+                except IndexError:
+                    print("Associated Receivers entity not found in Workspace.")
```

Comment:
**suggestion (logging):** use the [logging module]|https://docs.python.org/3/library/logging.html]. `logging.error()` or `logging.warning()`
well, if not in place yet for geoh5py, to be addressed later. That would take a new technical issue, replacing print statements by logging

Reply:
Will be addressed in GEOPY-444